mets_adapter/mets_adapter/src/main/scala/uk/ac/wellcome/mets_adapter/services/MetsAdapterWorkerService.scala

```diff
@@ -14,43 +15,116 @@ import uk.ac.wellcome.mets_adapter.models._
+  metsStore: MetsStore,
   concurrentConnections: Int = 6)(implicit ec: ExecutionContext)
-    extends Runnable {
+    extends Runnable
+    with Logging {
+
+  type Result[T] = Either[Throwable, T]
 
   val className = this.getClass.getSimpleName
 
   def run(): Future[Done] =
-    sqsStream.runStream(className, source => {
-      source
-        .via(retrieveBag)
-        .via(getMetsLocation)
-        .via(publishMetsLocation)
-    })
-
-  def retrieveBag: Flow[(SQSMessage, StorageUpdate), (SQSMessage, Bag), _] =
-    Flow[(SQSMessage, StorageUpdate)]
+    msgStream.runStream(
+      className,
+      source => {
+        source
+          .map { case (msg, update) => (Context(msg, update.bagId), update) }
+          .via(retrieveBag)
+          .via(parseMetsData)
+          .via(filterMetsData)
+          .via(publishMetsData)
+          .via(storeMetsData)
+          .map { case (Context(msg, _), _) => msg }
+      }
+    )
+
+  def retrieveBag =
+    Flow[(Context, IngestUpdate)]
       .mapAsync(concurrentConnections) {
-        case (msg, update) => bagRetriever.getBag(update).map(bag => (msg, bag))
+        case (ctx, update) =>
+          bagRetriever
+            .getBag(update)
+            .transform(result => Success((ctx, result.toEither)))
+      }
+      .via(catchErrors)
+
+  def parseMetsData =
+    Flow[(Context, Option[Bag])]
+      .mapWithContext { case (ctx, bag) => bag.metsData }
+
+  def filterMetsData =
+    Flow[(Context, Option[MetsData])]
+      .mapOptionalWithContext {
+        case (ctx, data) => metsStore.filterMetsData(ctx.bagId, data)
+      }
+
+  def publishMetsData =
+    Flow[(Context, Option[MetsData])]
+      .mapWithContext {
+        case (ctx, data) => msgSender.sendT(data).toEither.right.map(_ => data)
+      }
+
+  def storeMetsData =
+    Flow[(Context, Option[MetsData])]
+      .mapWithContext {
+        case (ctx, data) => metsStore.storeMetsData(ctx.bagId, data)
       }
-      .collect { case (msg, Some(bag)) => (msg, bag) }
 
-  def getMetsLocation: Flow[(SQSMessage, Bag), (SQSMessage, MetsLocation), _] =
-    throw new NotImplementedError
+  /** Encapsulates context to pass along each akka-stream stage. Newer versions
+    *  of akka-streams have the asSourceWithContext/ asFlowWithContext idioms for
+    *  this purpose, which we can migrate to if the library is updated.
+    */
+  case class Context(msg: SQSMessage, bagId: String)
 
-  def publishMetsLocation
-    : Flow[(SQSMessage, MetsLocation), SQSMessage, NotUsed] =
-    Flow[(SQSMessage, MetsLocation)]
+  /** Allows mapping a flow with a function, where:
+    *  - Context is passed through.
+    *  - None values are ignored and passed to the next stage: this is required
+    *    for the SQS delete action to trigger at the end of the stream
+    *  - Any errors are caught and the message prevented from propagating downstream,
+    *    resulting in the message being put back on the queue / on the dlq.
+    */
```

Comment:
Do you think this is relevant here?
https://github.com/wellcometrust/catalogue/pull/259#discussion_r348358717

Reply:
These comments are superb!